File: src/leetcode/sliding_window/subarrays_with_avg_geq_threshold.py
from typing import List


class Solution:
    def numOfSubarrays(self, arr: List[int], k: int, threshold: int) -> int:
        """
        Count sub-arrays of size k whose average is >= threshold.

        Args:
            arr: The list of integers to search.
            k: The fixed sub-array length.
            threshold: The minimum average a window must reach to count.

        Returns:
            The number of length-k windows whose average is >= threshold.
        """
        target = threshold * k  # compare sums directly, no float division
        window_sum = 0
        count = 0

        # Single unified loop: every index is treated the same way, instead
        # of special-casing the first window before the loop starts.
        for right in range(len(arr)):
            # Always add the incoming element first. This runs on every
            # iteration, including the first k-1 ones where the window
            # hasn't "filled up" yet.
            window_sum += arr[right]

            # Once the window would grow past size k, evict the element
            # that's k positions behind `right` -- that's the one that no
            # longer belongs in a size-k window ending at `right`.
            if right >= k:
                window_sum -= arr[right - k]

            # Only check the sum once the window has actually reached size
            # k for the first time (right == k - 1) or later. Before that,
            # window_sum is a partial sum, not a valid k-length window.
            if right >= k - 1 and window_sum >= target:
                count += 1

        return count


if __name__ == "__main__":
    solution = Solution()

    # Expected: 3 ([2,5,5,7,8] windows: [2,5,5]->4, [5,5,7]->5.67, [5,7,8]->6.67, ...)
    print(solution.numOfSubarrays([2, 2, 2, 2, 5, 5, 5, 8], 3, 4))

    # Expected: 6
    print(solution.numOfSubarrays([11, 13, 17, 23, 29, 31, 7, 5, 2, 3], 3, 5))

    # Expected: 5 (k=1, threshold=0 -> every single-element window passes)
    print(solution.numOfSubarrays([1, 1, 1, 1, 1], 1, 0))

    # Expected: 0 (no window reaches the threshold)
    print(solution.numOfSubarrays([1, 1, 1, 1, 1], 5, 5))

File: src/leetcode/sliding_window/subarrays_with_avg_geq_threshold.jl
"""
    numOfSubarrays(arr::Vector{Int}, k::Int, threshold::Int)::Int

Count sub-arrays of size `k` whose average is >= `threshold`.

Arguments:
  arr       -- the vector of integers to search
  k         -- the fixed sub-array length
  threshold -- the minimum average a window must reach to count

Returns the number of length-k windows whose average is >= threshold.
"""
function numOfSubarrays(arr::Vector{Int}, k::Int, threshold::Int)::Int
    # Hint: comparing `window_sum / k >= threshold` drags floats (or, with
    # `//`, allocating Rationals) into the loop. Scale the threshold up once
    # instead so every comparison stays in Int.
    target = threshold * k

    window_sum = 0
    count = 0

    # `left` trails `right`; the live window is always arr[left:right], so its
    # size is `right - left + 1` -- no offset arithmetic to second-guess.
    left = 1

    for right in 1:length(arr)
        # Extend: every element enters the window exactly once.
        window_sum += arr[right]

        # Shrink: the window overgrew, so drop the trailing element. A
        # fixed-size window can only overshoot by one per step, so this is an
        # `if`; a variable-size window would need a `while` here instead.
        if right - left + 1 > k
            window_sum -= arr[left]
            left += 1
        end

        # Count: a sibling of the shrink, not nested inside it. The first full
        # window (right == k) is built entirely by extends with zero shrinks,
        # so gating this on the shrink would skip it -- counting happens on
        # n-k+1 iterations, shrinking on only n-k of them.
        if right - left + 1 == k && window_sum >= target
            count += 1
        end
    end

    return count
end


if abspath(PROGRAM_FILE) == @__FILE__
    # Expected: 3 (qualifying windows are [2,5,5]->4, [5,5,5]->5, [5,5,8]->6;
    # the four windows before them average 2 or 3, below the threshold of 4)
    println(numOfSubarrays([2, 2, 2, 2, 5, 5, 5, 8], 3, 4))

    # Expected: 6
    println(numOfSubarrays([11, 13, 17, 23, 29, 31, 7, 5, 2, 3], 3, 5))

    # Expected: 5 (k=1, threshold=0 -> every single-element window passes)
    println(numOfSubarrays([1, 1, 1, 1, 1], 1, 0))

    # Expected: 0 (no window reaches the threshold)
    println(numOfSubarrays([1, 1, 1, 1, 1], 5, 5))

    # Expected: 1 (single window spanning the whole array, average exactly 3)
    println(numOfSubarrays([1, 2, 3, 4, 5], 5, 3))
end
